### ISSUE TITLE
Compiling Proton howto needed for Ubuntu 18.04 or further expanded

### ISSUE THREAD

**Issue by roadapathy** [ValveSoftware/Proton#1756]
# Feature Request
A more exhaustive howto for compiling Proton- especially for Ubuntu 18.04, one of the most popular distros for running Steam and Steam Play.

## I confirm:
- [x] that I haven't found another request for this feature.
- [x] that I have checked whether there are updates for my system available that
      contain this feature already.

## Description 
Ideally, a step-by-step cut-and-paste howto for Ubuntu 18.04 for users to ADD a custom compiled version of the latest Proton to their current Steam game client. IE., allowing for Proton 3.7, Proton Beta, and Proton Custom.

## Justification [optional] 
People need to know how to do this to increase performance for their specific CPU.

## Risks [optional]
People who aren't really developers asking you too many questions about the build process- which you may have already.

## References [optional] 
I'm sorry to be "that guy" but imagine how awesome this would be when compiled for a Ryzen 7. There are no howtos online other than this here. Could use some help.

**Comment by ghost** [comment]
If a person is going to get that technical about it, I'd say how it is now is appropriate and be content with that.

Building yourself has too many cons and too little pros. Just get playing instead. There's a certain point where squeezing out more performance is a effort not worth doing as you then spend all your time trying to get x fps more instead of actually playing the game. so that the game becomes not the game.

Perhaps it would be different if it was say 30fps with Steam and 60fps build yourself, but yeah, don't see that likely lol.

Such as with Wine  https://forum.winehq.org/viewtopic.php?f=2&t=28470
"Using native build flags (i.e. optimisations for your particular CPU archiecture) will make only a very marginal difference"

I say just play a game instead.

**Comment by tingvarsson** [comment]
Would probably be valuable to state which parts of the howto needs to be improved. What parts are you having difficulties with?

**Comment by Draycia** [comment]
Well, it would be nice to be able to manually apply patches that allow games and to work and to enable modding for them. Especially when there are no proton releases containing them, official or 3rd party.

**Comment by roadapathy** [comment]
I'm not trying to modify anything. I'm trying to pass options to the compiler to produce a faster binary for my system. Especially -O3 -march=native. Not trying to modify anything. I already compile the Kernel and my game FPS has increased dramatically. For an old PC, it could be the difference between playing No Man's Sky and not playing it.

**Comment by roadapathy** [comment]
One thing that would be great and help a lot of people is to take a baseline Ubuntu 18.04 and tell us all the dependencies. Post the command:

apt-get install x y z etc

From there, the Arch, Debian, Mint people could all figure out easily what is needed without compiling first and checking the log. Well, I take that back. They would still have to do it but maybe only once instead of a dozen times. :-)

**Comment by roadapathy** [comment]
> Building yourself has too many cons and too little pros. Just get playing instead. There's a certain point where squeezing out more performance is a effort not worth doing as you then spend all your time trying to get x fps more instead of actually playing the game. so that the game becomes not the game.

I think you had better go back to Windows because you don't really seem to understand why we use Linux. I compile libraries, Kernels, even entire games because I can- In Windows, you generally can't.

**Comment by aeikum** [comment]
Please don't be rude.

The instructions provided are for Debian, including a list of packages needed. Ubuntu is Debian-based, so the requirements should be similar. If you are building outside of the Steam runtime and you are missing Wine build dependencies, check out apt-get's build-dep verb and the Wine wiki for more help on building Wine.

If you have specific suggestions for improvements, we're open to them, but I don't think we're interested in providing a copy-paste set of instructions. Users should understand basic package management and software building steps before attempting to build Proton.

**Comment by roadapathy** [comment]
> If you have specific suggestions for improvements, we're open to them, but I don't think we're interested in providing a copy-paste set of instructions. Users should understand basic package management and software building steps before attempting to build Proton.

I understand. When I was in C class, and I'm not a developer, my professor said developers hate documenting their work. So I understand that you don't want to do it- but sadly, we can't do it for you because we're not developers. You see the catch 22 there? Ubuntu is one of the most widely used distros for Steam on Linux. 

https://store.steampowered.com/hwsurvey/Steam-Hardware-Software-Survey-Welcome-to-Steam?platform=linux

Currently, there are no other HowTos. I have looked around trying to outsource this wiki. No luck.

I have compiled Wine in the past but I haven't been able to since 2012 and I think it's because of the 32bit and 64bit libraries needed (both). I can only compile a 64bit Wine.

Why don't you just help us out a little bit?

**Comment by ghost** [comment]
> > Building yourself has too many cons and too little pros. Just get playing instead. There's a certain point where squeezing out more performance is a effort not worth doing as you then spend all your time trying to get x fps more instead of actually playing the game. so that the game becomes not the game.
> 
> I think you had better go back to Windows because you don't really seem to understand why we use Linux. I compile libraries, Kernels, even entire games because I can- In Windows, you generally can't.

I use Linux because I can. That's the whole point to your point. You can also build it yourself without a step-by-step guide because you use Linux and because you can. No one's stopping you, but you because this is Linux.

**Comment by roadapathy** [comment]
> I use Linux because I can. That's the whole point to your point. You can also build it yourself without a step-by-step guide because you use Linux and because you can. No one's stopping you, but you because this is Linux.

My kind and polite advice is this: Please do not tell people "Building yourself has too many cons and too little pros" to any Linux user. This isn't our part of our ethos. Source code isn't provided just for the heck of it. Do you see how you came off as very rude when you said that?

Additionally, you said "There's a certain point where squeezing out more performance is a effort not worth doing as you then spend all your time trying to get x fps more instead of actually playing the game" to a person who enjoys compiling from source and is here asking for some clarification. Do you kind of see how that came across as rude?

Compiling from source can:
A. Add more features using newer builds
B. Fix more bugs using newer builds
C. Add some performance due to CPU optimized code, like -O and -march
D. It's fun and that's something Linux people like to do
E. It's something to share with Windows people to show how versatile the OS is

**Comment by roadapathy** [comment]
The first time I compiled my Kernel, it felt like I upgraded the CPU. I remember compiling for my specific CPU. Back then, it made a huge difference. Today there are more variables but it still has an effect and sometimes very big. My gf's system was having trouble running her Linux game. I downloaded the newer Kernel, compiled it, added my fine-tuning, and then ran her game. It no longer stutters. I actually couldn't believe how much faster it was. It's not always this way and maybe for Proton it won't be.... but maybe it will and I feel that I must find out. I also have to report back to my game group. There's 400+ of us waiting to find out.

**Comment by ghost** [comment]
The funny thing is you don't even know how long I've been using Linux for, how many distro's I use, how much or little my Linux knowledge is and you're telling me how Linux people are. So, this is my final comment to any of your posts.

Why don't you try building it and working through it, detailing it here, instead of having someone else do it for you. That's a real Linux user.

**Comment by roadapathy** [comment]
> The funny thing is you don't even know how long I've been using Linux for, how many distro's I use, how much or little my Linux knowledge is and you're telling me how Linux people are. So, this is my final comment to any of your posts.

Why would I need to know any of that? I'm only here to help the project in whatever small way that I can so that more people can use this better. If the build instructions are further expanded and it helps others build who normally couldn't then it has improved the project. I'm just going to use whatever I can get to build the project, create a distro-specific howto, and share it with the world like I usually do. Nothing wrong with that.

**Comment by roadapathy** [comment]
Here's a cool trick. I extracted the Nvidia drivers and then did this:

grep -r CXXFLAGS
grep -r CFLAGS

and also

grep -r CFLAGS | grep O2

I found that Nvidia drivers will compile the GPU Kernel module. I changed -O2 to -O3 and added -march=native. I can't tell that it worked but it was worth a shot. Same with compiling Wine.

**Comment by kisak-valve** [comment]
Hello @roadapathy, this issue tracker is for actionable issues with Proton. If you have specific issues with the documentation, please open a new issue report for that.

Closing as this feature request is too generalized to be actionable.
